### ISSUE TITLE
DAT (M. A. M. E. 0.37b5) that crashes JRomManager when importing.

### ISSUE THREAD

**Issue by optyfr** [optyfr/JRomManager#39]
Hey optyfr, I'm adding you another DAT (M. A. M. E. 0.37b5) that crashes JRomManager when importing. Could you please see it and check if it's fixable? Thanks!
[mame037b5_2nd.xml.zip](https://github.com/optyfr/JRomManager/files/5360254/mame037b5_2nd.xml.zip)

_Originally posted by @cinnamonmatexfce in https://github.com/optyfr/JRomManager/issues/26#issuecomment-706615492_

**Comment by optyfr** [comment]
The trouble is starting at line 2846 where size for roms of game arkbl3 are equals to "-" !!!
Size is a required field **and** must be a valid number, even if I replace with 0, size is an important part of the scan and you absolutely need a valid size corresponding to the real size of the rom file and preferably a CRC/MD5/SHA1 to verify data inside file, the only situation where we can assume it's 0 size is when checksum is also 0 or status="nodump"

All that to say that if I replace these values with zero, it will permit to load the dat, **but** you won't be able to build/fix these games with zero sized roms and non zeroed checksum, I'm not even sure that's worth it to modify the scan algorithm just for those old and incomplete dat files

I'm pretty sure that I already got a MAME 0.37b5 dat files that loaded fine, how did you got this one? and what is the "_2nd" appended?

**Comment by optyfr** [comment]
I will at least avoid it to crash, but I won't let it load for the time being...

**Comment by cinnamonmatexfce** [comment]
> how did you got this one? and what is the "_2nd" appended?


It's an "unoffical" XML, exported as search results by (my) filters.
See: http://adb.arcadeitalia.net/lista_mame.php

![xml_exporting_search_results_adb_arcadeitalia](https://user-images.githubusercontent.com/65762930/95679504-90c88f00-0bd3-11eb-9a04-2d2028a3252c.jpg)

**Comment by cinnamonmatexfce** [comment]
Indeed, the **"official" 0.37b5 DAT works fine**.
Attachments: good DAT file.
[MAME 0.37b5.zip](https://github.com/optyfr/JRomManager/files/5361132/MAME.0.37b5.zip)

**Comment by optyfr** [comment]
Ok, so was arkbl3 supposed to be removed after filtering?

**Comment by cinnamonmatexfce** [comment]
Nope. That export ("_2nd") is the initial test with **ALL** sets enabled from year 2000 release (without filtering).

BTW, other XML filtered exports (from other M. A. M. E. versions) works good in JRomManager.

**Comment by optyfr** [comment]
in fact arkbl3 does not exist in the official 0.37b5
After trying with other wrong games, seems that all games with roms of size="-" are not in the "official" one

**Comment by optyfr** [comment]
can we access the source xml used (even if you outputted with all enabled)?

**Comment by cinnamonmatexfce** [comment]
Don't know... I'll try to contact the author of http://adb.arcadeitalia.net/lista_mame.php and ask him he he could help or at least give the infos you need.

Some public infos are here: http://adb.arcadeitalia.net/ringraziamenti.php

**Comment by optyfr** [comment]
I don't know how they are building that filtered xml, but that's not from 0.37b5 which is available here : http://adb.arcadeitalia.net/download.php?versione_mame=98
Another particularity with these odd games, is that there is no useful description, example : for game id "arkbl3", description is "arkbl3", instead of "Arkanoid (Japanese Bootleg 3)"

**Comment by cinnamonmatexfce** [comment]
Mmm... Interesting... `mame=98`? Isn't the 0,37b5 supposed to be 0.41?

**Comment by optyfr** [comment]
98 is not a version but just an index count from the start including beta and subversions

possible that it's extracted from a big multi-versions SQL database on the backend server, and that there is some mistakes or incomplete informations in that database, what I see also in the history of those games is that they were all released for versions older than 0.37b5 but were not officially made available for that specific version (don't know why), then they were renamed later in version above 0.100

**Comment by optyfr** [comment]
I think that the best you could do is fill the form at http://adb.arcadeitalia.net/contatti.php to report error about extra games when generating a dat for 0.37b5 (in comparison to the official one), and that those extra games are all with rom size set to "-" and with no useful title...
In the mean time I will try to "bullet proof" the profile loading in jrommanager

**Comment by cinnamonmatexfce** [comment]
Sure! As soon as they get me answer back I'll let you know.
Thanks!

**Comment by phweda** [comment]
@optyfr are you validating the incoming DATs? If not find the XSD in MFM.

[https://github.com/phweda/MFM/blob/master/Settings/datafile2018.xsd](url)

**Comment by optyfr** [comment]
Hello @phweda, no I disabled validation because I already encountered some very bad datfiles before (some starting with \<datafile\> but closing with \</mame\> and viceversa or with no closing tag at all). 

But thanks for pointing me out your xsd file. The problem here is that it's telling me that it is a string but not telling me what it can contain and what to do with "funny sizes", I already knew that it could contain hex numbers and not only base 10 plain integers, but now there is also "-" which doesn't tell me exactly how I should handle this information : is it similar to a baddump rom? a nodump rom? or a legit rom but with unknown size information so that I should take care of CRC only without taking into account the size while scanning/fixing? I would tend to think of the last hypothesis probably, even if it's strange!

That's for the size="-" problematic... now I discovered that http://adb.arcadeitalia.net/ is generating datfiles supposed to be for MAME 2000 aka Mame 0.37b5, but with extra games to the official complete datfile, and those extra games are all with those funny sizes and with no useful \<description\>

**Comment by phweda** [comment]
Problem is the DTD is ~20 years old. No defined ATTR value types. So:
```
<!ELEMENT rom EMPTY>
	<!ATTLIST rom name CDATA #REQUIRED>
	<!ATTLIST rom bios CDATA #IMPLIED>
	<!ATTLIST rom size CDATA #REQUIRED>
	<!ATTLIST rom crc CDATA #IMPLIED>
```

When I created the XSD left these as is so CDATA becomes xs:string. Am sure that for core MAME **size** is always numeric. No Hex etc.

IMHO you should reject these DATs. There are a lot of these 'bad' DAT files floating out there. _Mostly where a human has tried to modify an existing DAT by hand._ Report out it is bad and handle as you did here advising what is wrong and to check with the source. Ignoring size will not work as the emulator will check that. Any 'legit' ROM has a known size(or 0?) that is in the source code. Again, in core MAME, all ROMs have a numeric size including Baddump and Nodump.

Suggestion knowing your desire and possible variances in emulator derivatives. Warn it is _bad_(your definition of that). Skip and report out which Machines have bad/incomplete entries. Additionally you could ask if user wants to continue with the good entries only.

FYI have never tested MFM with derivatives. It will work if they follow the MAME <ins>standard<ins>.

Regards Phweda

**Comment by motoschifo** [comment]
Hi, I'm the author of ADB website and the export procedure.
I have no issue with extra roms, I have tried various exports but the results are always corrects (i.e. no romset arkbl3 is on the file).

Xml file for release 0.37b5 of Mame is not available, this is created automatically using the original list of roms. No extra roms can be exported if it's selected the 0.37b5 in the export dialog, and I have added a 2nd check when "export a complete set" is selected. This test is redundant, but can be useful in particolar situations.

I will investigate about that.

IMHO the bad dat should rejected because this is an error, then I will correct and change the exports procedure if I'll will find issues.

**Comment by optyfr** [comment]
Hi @motoschifo regarding the above screenshot from @cinnamonmatexfce do you know how to access to this page. I can't find it on the web site...

**Comment by optyfr** [comment]
OK I think I found it, as is without searching/filtering on the main page, the exported xml is correct... so the question to @cinnamonmatexfce is what was your selected filters before choosing "Export the results"?

**Comment by motoschifo** [comment]
The url is [http://adb.arcadeitalia.net/?search=mame](http://adb.arcadeitalia.net/?search=mame), then you need to press the button OPTION and EXPORT THE SEARCH.
I made small changes on that page adding a double check of output romsets list.
I can't reproduce the original issue, but the option to filter the data by emulator release on the export dialog boxes is fundamental for dat and xml files.
Without that, the output can containts extra roms.
This is intentionally because I can reach the list by different path and select different format.

If you find any issues, please report them so I can reproduce & fix.

**Comment by cinnamonmatexfce** [comment]
> OK I think I found it, as is without searching/filtering on the main page, the exported xml is correct... so the question to @cinnamonmatexfce is what was your selected filters before choosing "Export the results"?

Just (almost) like you do. Only thing I **unchecked** in the MAIN PAGE (not inside additional filters) was "LATEST MAME RELEASE (0.225)". No filters was selected at all; in the EXPORT sub-window I checked: "EXPORT COMPLETE SET"  and "ADD CLONES OF ALL PARENT"

**Comment by optyfr** [comment]
@motoschifo is that the result of the double check?
![image](https://user-images.githubusercontent.com/17027109/96903826-9b6a0a80-1496-11eb-87e9-e4e26dc3a6f5.png)

**Comment by motoschifo** [comment]
Of course it is not, this is a general error (ex. timeout of web server sometime generate this kind of anomaly).

When you do exports, and there are no errors on server side (I see no error in the logs), the file is always generated. If you reload the page, after the OPTIONS button, you should find a link to your last export.

How did you get this error? Can you reproduce it?

**Comment by optyfr** [comment]
Ok, I reproduced it twice in a row when reporting previously, while trying the same export as @cinnamonmatexfce... and I reproduced it again just now! this happen relatively fast (at least faster than when it works)
This also happen when checking again "LATEST MAME RELEASE (0.225)" back, all when trying to export 0.37b5 (0.41)

**Comment by motoschifo** [comment]
On server I have no error, so I have done tests with different users/browsers, but I receive no errors.
The server is very busy, there are a lots of connections, may be a kind of timeout issue.

I will try again tomorrow.

**Comment by optyfr** [comment]
@motoschifo It worked this morning, and I can confirm that the size="-" is not there anymore and the dataset is loading entirely... \o/
@cinnamonmatexfce the extra check on my side will be available in 2.5.0 beta 2 soon, it won't load invalid dat without crashing, and also will tell to the user what/where (according type of error) was the problem in the dat file...
Thanks for the help
